optuna/_warnings.py

```diff
@@ -0,0 +1,42 @@
+from __future__ import annotations
+
+import os
+from pathlib import Path
+import sys
+import warnings
+from warnings import catch_warnings
+from warnings import filterwarnings
+from warnings import simplefilter
+
+
+_OPTUNA_MODULE_ROOT: str = (str(Path(__file__).resolve().parent) + os.sep).casefold()
+
+
+def find_stacklevel() -> int:
+    level = 1
+    try:
+        while True:
+            if (
+                not getattr(sys._getframe(level).f_code, "co_filename", "")
+                .casefold()
+                .startswith(_OPTUNA_MODULE_ROOT)
+            ):
+                return level
+            level += 1
+    except ValueError:
+        return level
```

Comment:
How about using `inspect` instead of `sys._getframe`, since `sys._getframe` is internal function and not officially supported?
This function can be rewritten using inspect as follows:
```suggestion
def find_stacklevel() -> int:
    level = 1
    frame = inspect.currentframe()
    while frame is not None:
        frame = frame.f_back
        filename = frame and frame.f_code and frame.f_code.co_filename
        if not filename or not filename.casefold().startswith(_OPTUNA_MODULE_ROOT):
            return level
        level += 1
    return level
```

Reply:
Okay, if we don't pursue this functionality on Python 3.11 or earlier, it would also be acceptable.